docker_images/peft/app/pipelines/text_generation.py

```diff
@@ -0,0 +1,81 @@
+import logging
+import os
+
+import torch
+from app import idle, timing
+from app.pipelines import Pipeline
+from huggingface_hub import model_info
+from peft import PeftModel
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+
+logger = logging.getLogger(__name__)
+
+
+class TextGenerationPipeline(Pipeline):
+    def __init__(self, model_id: str):
+        use_auth_token = os.getenv("HF_API_TOKEN")
+        model_data = model_info(model_id, token=use_auth_token)
+        config_dict = model_data.config["peft"]
+
+        if config_dict:
+            base_model_id = config_dict["base_model_name"]
+            if base_model_id:
+                self.tokenizer = AutoTokenizer.from_pretrained(base_model_id)
+                model = AutoModelForCausalLM.from_pretrained(
+                    base_model_id, device_map="auto"
+                )
+                # wrap base model with peft
+                self.model = PeftModel.from_pretrained(model, model_id)
+            else:
+                raise ValueError("There's no base model ID in configuration file.")
+        else:
+            raise ValueError("Config file for this model does not exist or is invalid.")
+
+    def __call__(self, inputs: str, **kwargs) -> str:
+        """
+        Args:
+            inputs (:obj:`str`):
+                a string for text to be completed
+        Returns:
+            A string of completed text.
+        """
+        if idle.UNLOAD_IDLE:
+            with idle.request_witnesses():
+                self._model_to_gpu()
```

Comment:
@Narsil do we want the GPU-handling? Except for `diffusers`, I think it's safe to assume CPU-only for free inference API no?

Reply:
For now yes. but if the code is GPU aware that's less work for when/if we enable it.

The only downside is code complexity.